### PR TITLE
Document new experimental ingestion streaming APIs

### PR DESCRIPTION
**Pull request by reta**
### Description
Document new experimental ingestion streaming APIs

### Issues Resolved
Closes https://github.com/opensearch-project/documentation-website/issues/8111

### Version
`2.17.0`

### Frontend features
_If you're submitting documentation for an OpenSearch Dashboards feature, add a video that shows how a user will interact with the UI step by step. A voiceover is optional._ 

### Checklist
- [X] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kolchfa-aws** [comment]
Thank you, @reta! I made some edits and pushed into your PR. Please verify that everything is still technically accurate. Also, I changed the bulk request to curl so please verify that the request is formatted correctly. Will this bulk request work in Dev Tools of OS Dashboards? (not sure if it'll work with the 2 headers).

**Comment by kolchfa-aws** [comment]
Also, please provide an issue where the users can track the progress of the feature. Or we can just direct them to the OS forum.

File: _api-reference/document-apis/bulk-streaming.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+nav_order: 25
+redirect_from:
+ - /opensearch/rest-api/document-apis/bulk/streaming/
+---
+
+# Streaming bulk
+**Introduced 2.17.0**
+{: .label .label-purple }
+
+This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://example.issue.link).
```
Comment (kolchfa-aws) [inline _api-reference/document-apis/bulk-streaming.md:14]: Please provide the issue # here.

**Review by kolchfa-aws — APPROVED** [review]
Thank you, @reta! I think we can still keep the `json` highlighting on the curl code block. Technically, it's not JSON, but I like the way the curly braces are matched and the field names are all the same color :) Applying "Done but waiting to merge" label to put this in the 2.17 editorial queue starting 9/11. This PR will be merged before the release.

**Comment by reta** [comment]
> Thank you, @reta! I think we can still keep the `json` highlighting on the curl code block.

Got it, brought it back, thanks @kolchfa-aws !

File: _api-reference/document-apis/bulk-streaming.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+---
+
+# Streaming bulk
+**Introduced 2.17.0**
+{: .label .label-purple }
+
+This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/9065).    
+{: .warning}
+
+The streaming bulk operation lets you add, update, or delete multiple documents in by streaming the request and getting the results as streaming response. In comparison to the traditional [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/), streaming ingestion eliminates the need to guess the batch size (which is affected by the cluster operational state at any given time) and naturally applies the backpressure between many clients and the cluster. The streaming works over HTTP/2 or HTTP/1.1 (using chunked transfer encoding), depending on the capabilities of the clients and the cluster.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:17]:
```suggestion
The streaming bulk operation lets you add, update, or delete multiple documents by streaming the request and getting the results as a streaming response. In comparison to the traditional [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/), streaming ingestion eliminates the need to estimate the batch size (which is affected by the cluster operational state at any given time) and naturally applies backpressure between many clients and the cluster. The streaming works over HTTP/2 or HTTP/1.1 (using chunked transfer encoding), depending on the capabilities of the clients and the cluster.
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+## Path and HTTP methods
+
+```json
+POST _bulk/stream
+POST <index>/_bulk/stream
+```
+
+If you specify the index in the path, then you don't need to include it in the [request body chunks]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/#request-body).
+
+OpenSearch also accepts PUT requests to the `_bulk/stream` path, but we highly recommend using POST. The accepted usage of PUT---adding or replacing a single resource at a given path---doesn't make sense for streaming bulk requests.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:31]: "on" a given path?

File: _api-reference/document-apis/bulk-streaming.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+
+## Query parameters
+
+The following table lists the available query parameters. All query parameters are optional.
+
+Parameter | Data type | Description
+:--- | :--- | :---
+`pipeline` | String | The pipeline ID for preprocessing documents.
+`refresh` | Enum | Whether to refresh the affected shards after performing the indexing operations. Default is `false`. `true` makes the changes show up in search results immediately, but hurts cluster performance. `wait_for` waits for a refresh. Requests take longer to return, but cluster performance doesn't suffer.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:42]:
```suggestion
`refresh` | Enum | Whether to refresh the affected shards after performing the indexing operations. Default is `false`. `true` causes the changes show up in search results immediately but degrades cluster performance. `wait_for` waits for a refresh. Requests take longer to return, but cluster performance isn't degraded.
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+`require_alias` | Boolean | Set to `true` to require that all actions target an index alias rather than an index. Default is `false`.
+`routing` | String | Routes the request to the specified shard.
+`timeout` | Time | How long to wait for the request to return. Default `1m`.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:45]:
```suggestion
`timeout` | Time | How long to wait for the request to return. Default is `1m`.
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+`type` | String | (Deprecated) The default document type for documents that don't specify a type. Default is `_doc`. We highly recommend ignoring this parameter and using a type of `_doc` for all indexes.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:46]:
```suggestion
`type` | String | (Deprecated) The default document type for documents that don't specify a type. Default is `_doc`. We highly recommend ignoring this parameter and using the `_doc` type for all indexes.
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+`wait_for_active_shards` | String | Specifies the number of active shards that must be available before OpenSearch processes the bulk request. Default is 1 (only the primary shard). Set to `all` or a positive integer. Values greater than 1 require replicas. For example, if you specify a value of 3, the index must have two replicas distributed across two additional nodes for the request to succeed.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:47]:
```suggestion
`wait_for_active_shards` | String | Specifies the number of active shards that must be available before OpenSearch processes the bulk request. Default is `1` (only the primary shard). Set to `all` or a positive integer. Values greater than 1 require replicas. For example, if you specify a value of 3, then the index must have 2 replicas distributed across 2 additional nodes in order for the request to succeed.
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+`batch_interval` | Time | Specifies how long bulk operations should be accumulated into batch before sending over to data nodes.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:48]:
```suggestion
`batch_interval` | Time | Specifies for how long bulk operations should be accumulated into a batch before sending the batch to data nodes.
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+`batch_size` | Time | Specifies how many bulk operations should be accumulated into batch before sending over to data nodes. Default `1`.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:49]:
```suggestion
`batch_size` | Time | Specifies how many bulk operations should be accumulated into a batch before sending the batch to data nodes. Default is `1`.
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+{% comment %}_source | List | asdf
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:50]: Should this and the following two rows be here?

File: _api-reference/document-apis/bulk-streaming.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+`_source_excludes` | list | asdf
+`_source_includes` | list | asdf{% endcomment %}
+
+## Request body
+
+The streaming bulk API request body is fully compatible with the [Bulk API request body]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/#request-body), where each bulk operation (create/index/update/delete) is sent as a separate chunk.
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:56]:
```suggestion
The Streaming Bulk API request body is fully compatible with the [Bulk API request body]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/#request-body), where each bulk operation (create/index/update/delete) is sent as a separate chunk.  
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+{ "title": "Prisoners", "year": 2013 }
+{ "update": { "_index": "movies", "_id": "tt0816711" } }
+{ "doc" : { "title": "World War Z" } }
+'
+```
+{% include copy.html %}
+
+## Example response
+
+Depending on the batch settings, each streamed response chunk may report the results of one or many (batch) bulk operations, for example for the preceding request with no batching (default), the streaming response may look like this:
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:75]:
```suggestion
Depending on the batch settings, each streamed response chunk may report the results of one or many (batch) bulk operations. For example, for the preceding request with no batching (default), the streaming response may appear as follows:
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+`routing` | String | Routes the request to the specified shard.
+`timeout` | Time | How long to wait for the request to return. Default `1m`.
+`type` | String | (Deprecated) The default document type for documents that don't specify a type. Default is `_doc`. We highly recommend ignoring this parameter and using a type of `_doc` for all indexes.
+`wait_for_active_shards` | String | Specifies the number of active shards that must be available before OpenSearch processes the bulk request. Default is 1 (only the primary shard). Set to `all` or a positive integer. Values greater than 1 require replicas. For example, if you specify a value of 3, the index must have two replicas distributed across two additional nodes for the request to succeed.
+`batch_interval` | Time | Specifies how long bulk operations should be accumulated into batch before sending over to data nodes.
+`batch_size` | Time | Specifies how many bulk operations should be accumulated into batch before sending over to data nodes. Default `1`.
+{% comment %}_source | List | asdf
+`_source_excludes` | list | asdf
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:51]:
```suggestion
`_source_excludes` | List | asdf
```

File: _api-reference/document-apis/bulk-streaming.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+`_source_includes` | list | asdf{% endcomment %}
```
Comment (natebower) [inline _api-reference/document-apis/bulk-streaming.md:52]:
```suggestion
`_source_includes` | List | asdf{% endcomment %}
```

**Review by natebower — COMMENTED** [review]
@reta @kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!

File: _api-reference/document-apis/bulk-streaming.md (hunk 4 above)
Comment (reta) [reply _api-reference/document-apis/bulk-streaming.md:42]: So this paragraph is taken as is from existing docs on Bulk API [1], I am fine changing it but it should be done both places. Sounds good to you?

[1] https://opensearch.org/docs/latest/api-reference/document-apis/bulk/#url-parameters

File: _api-reference/document-apis/bulk-streaming.md (hunk 10 above)
Comment (reta) [reply _api-reference/document-apis/bulk-streaming.md:50]: They should (by API specs)

File: _api-reference/document-apis/bulk-streaming.md (hunk 4 above)
Comment (kolchfa-aws) [reply _api-reference/document-apis/bulk-streaming.md]: Yes, we can change in both places. I updated the suggestion so it flows nicer.

**Pull request merged**